### ISSUE TITLE
PTX inline assembly for using tensor cores

### ISSUE THREAD

**Issue by ihavnoid** [gcp/leela-zero#2036]
On #2007 I mentioned that I will attempt writing inline assembly to use the tensor cores.  Well, I think I got a functional version of hgemm that uses the wmmv instructions!

https://github.com/ihavnoid/hgemmtest

Currently it runs around 30ms on a Tesla v100 (tested on Google Compute Platform).  This is similar to what I got while running the leela-zero tuner, so it's hardly any improvement yet.  Not bad for the first implementation though. :)

```
(228/290) KWG=32 KWI=8 MDIMA=8 MDIMC=8 MWG=16 NDIMB=8 NDIMC=8 NWG=32 SA=1 SB=1 STRM=0 STRN=0 VWM=2 VWN=4 0.0317 ms (3716.1 GFLOPS)
```

Since this is only the first iteration, hopefully we can better tune the matrix multiplication code, and one day when the assembly version performs much better we can attempt to merge it into leela-zero.  I need some help from people who has better expertise on fine-tuning these kind of kernels.

**Comment by ihavnoid** [comment]
Plus, I would like to have some people test the code on how it performs on an Turing GPU

**Comment by ihavnoid** [comment]
Correction : 30us.  Not 30 ms.  Got a bit more tuning (m32n8k16 instead of m16n16k16) and now I am at 25us. :)

**Comment by Ttl** [comment]
Very interesting. It might be hard to utilize the hardware fully at batch size of 1. Batching in Winograd increases the n-dimension in sgemm (n = batch_size * tiles and tiles = 25 for F(4x4, 3x3)) allowing use of bigger tiles.

Currently transformations take about 30 - 40% of runtime so doubling the sgemm speed results in about 50% nps increase. Cudnn actually performs better using im2col + sgemm instead of Winograd at low batch sizes even though it's theoretically more work, but it probably uses tensor cores more effectively.

NVidia doesn't provide profiler for opencl anymore and the only way to profile opencl kernels is to compile them with cuda. It isn't as hard it seems, kernels usually work just by making some defines (For example: https://github.com/CNugteren/CLBlast/blob/master/src/kernels/opencl_to_cuda.h). Host code also needs to be written with cuda and there are examples for that too in CLBlast.

**Comment by sewim** [comment]
It is going to be happened? Interesting...
So basically, Leela zero will be released normal version and also Nvidia specific version?

**Comment by alreadydone** [comment]
I compiled with Visual Studio (project file included in https://github.com/alreadydone/hgemmtest) and when I run it I get the following output. Is this expected? Must the GPU has tensor cores?
<detail>

```
E:\hgemmtest\x64\Release>hgemmtest
Detected 2 OpenCL platforms.
Platform version: OpenCL 1.2
Platform profile: FULL_PROFILE
Platform name:    Intel(R) OpenCL
Platform vendor:  Intel(R) Corporation
Device ID:     0
Device name:   Intel(R) HD Graphics 4600
Device type:   GPU
Device vendor: Intel(R) Corporation
Device driver: 20.19.15.4703
Device speed:  1150 MHz
Device cores:  20 CU
Device score:  612
Device ID:     1
Device name:   Intel(R) Core(TM) i7-4712MQ CPU @ 2.30GHz
Device type:   CPU
Device vendor: Intel(R) Corporation
Device driver: 5.2.0.10094
Device speed:  2300 MHz
Device cores:  8 CU
Device score:  512
Platform version: OpenCL 1.2 CUDA 10.0.132
Platform profile: FULL_PROFILE
Platform name:    NVIDIA CUDA
Platform vendor:  NVIDIA Corporation
Device ID:     2
Device name:   GeForce 940M
Device type:   GPU
Device vendor: NVIDIA Corporation
Device driver: 416.34
Device speed:  1176 MHz
Device cores:  3 CU
Device score:  1112
Selected platform: NVIDIA CUDA
Selected device: GeForce 940M
with OpenCL 1.2 capability.
sourceCode :
Error building kernels:
```
</details>

**Comment by iteachcs** [comment]
I compiled and ran the program in Ubuntu 18.04 on a workstation with 2 P5000 and 1 Titan V.  The result is as follows:
<details>
<pre>
Detected 1 OpenCL platforms.
Platform version: OpenCL 1.2 CUDA 10.0.185
Platform profile: FULL_PROFILE
Platform name:    NVIDIA CUDA
Platform vendor:  NVIDIA Corporation
Device ID:     0
Device name:   TITAN V
Device type:   GPU
Device vendor: NVIDIA Corporation
Device driver: 410.73
Device speed:  1455 MHz
Device cores:  80 CU
Device score:  1112
Device ID:     1
Device name:   Quadro P5000
Device type:   GPU
Device vendor: NVIDIA Corporation
Device driver: 410.73
Device speed:  1733 MHz
Device cores:  20 CU
Device score:  1112
Device ID:     2
Device name:   Quadro P5000
Device type:   GPU
Device vendor: NVIDIA Corporation
Device driver: 410.73
Device speed:  1733 MHz
Device cores:  20 CU
Device score:  1112
Selected platform: NVIDIA CUDA
Selected device: TITAN V
with OpenCL 1.2 capability.
0.000075 0.085000
</pre>
</details>

**Comment by ihavnoid** [comment]
@alreadydone yes it will crash because it will refuse to compile the opencl code.

**Comment by ihavnoid** [comment]
The last two numbers are the average error and the runtime (ms).  Kernel tuning is not implemented yet so your result may vary.

**Comment by alreadydone** [comment]
Hmm... I didn't know that hgemm.cl needs to be put under the release folder that contains hgemmtest.exe. When I do it I get
`0.000296 1.400816`
and when I uncomment #define USE_TC I get the long output [output.txt](https://github.com/gcp/leela-zero/files/2609739/output.txt) (irrelevant though as the GPU doesn't have tensor cores)

**Comment by Yakago** [comment]
I have an RTX card and would like to help - however I don't know how to build this myself. (Don't have much experience with these kind of things)
I tried downloading VS2017 and build using Alreadydone's files - but I don't know what to do from there. hgemmtest.exe just writes 'OpenCL: clGetPlatformIDs' and then closes, so I assume I must have done something wrong at some point

**Comment by ihavnoid** [comment]
Copy the hgemm.cl file where hgemmtest.exe is, edit the file to uncomment where it defines USE_TC and try rerunning the binary.

Please copy and paste the full output here. :)

**Comment by alreadydone** [comment]
I already uncommented in my fork. OpenCL: clGetPlatformIDs' isn't an error I'd expect (I didn't get it when hgemm.cl is missing).

**Comment by Yakago** [comment]
I get
Detected 1 OpenCL platforms.
Platform version: OpenCL 1.2 CUDA 10.0.132
Platform profile: FULL_PROFILE
Platform name:    NVIDIA CUDA
Platform vendor:  NVIDIA Corporation
Device ID:     0
Device name:   GeForce RTX 2080 Ti
Device type:   GPU
Device vendor: NVIDIA Corporation
Device driver: 416.94
Device speed:  1650 MHz
Device cores:  68 CU
Device score:  1112
Selected platform: NVIDIA CUDA
Selected device: GeForce RTX 2080 Ti
with OpenCL 1.2 capability.
0.000422 0.043320

**Comment by ihavnoid** [comment]
Implemented a crude tuner and pushed it into repository.  This is what I got from a tesla V100 in a google compute platform:  (the winner changes every time I try again)
```
Winner :  -DMDIMC=1 -DNDIMC=1 -DMWG=2 -DNWG=2 -DKWG=1 -DSA=0 -DSB=0 -DVWM=1 -DVWN=1 0.000752 0.019712
```
20us per run is quite quick - will have a look if it can get shorter but probably not... :)

**Comment by ihavnoid** [comment]
Integrated the gemm code into leela-zero.  This is what I have:

https://github.com/ihavnoid/leela-zero/tree/tensorcore

Around 10% improvement on a v100.  As @Ttl mentioned we aren't going to improve a whole lot because xgemm is not the bottleneck (the in/out transformation is).  Can others with smaller consumer GPUs also test what you get?

**Comment by alreadydone** [comment]
`Tensor Core support: No.` on V100 and Titan V:
![image](https://user-images.githubusercontent.com/3064145/48965532-562cbd80-ef8d-11e8-9dfa-6f675b68a026.png)

With double GPUs:
![image](https://user-images.githubusercontent.com/3064145/48965538-65ac0680-ef8d-11e8-9f5a-a043fdcd9752.png)
![image](https://user-images.githubusercontent.com/3064145/48965541-6cd31480-ef8d-11e8-9b39-59d276f04000.png)
With single GPU:
![image](https://user-images.githubusercontent.com/3064145/48965544-7197c880-ef8d-11e8-8e79-da1e943b6344.png)

**Comment by pangafu** [comment]
@alreadydone @ihavnoid 

I get same result on TITAN V.
1 card TITAN V report there is no tensorcore
2 card TITAN V core dump

**Comment by ihavnoid** [comment]
@alreadydone I don't know - I am running on 410.48.  Should work on anything with CUDA 9.x (at least according to nvidia specs).

**Comment by alreadydone** [comment]
Updating driver to 415.13 solved problem for the V100.

**Comment by ihavnoid** [comment]
To check if there is anything wrong with the ptx assembly, I tested a 1024x1024 matrix multiplication with a batch size of 36.  This is 72giga floating point ops, which took 1.2ms.  Which means I was getting at least 60 tera floating point ops, which translates to 50% of tensor core perf.

The memory footprint should be 1024x1024x3x2x36 = 216MB.  Since v100 has 900GB/sec of memory bandwidth, it should take much less than 0.3msec to load the whole matrix to GPU.  Hence we aren't memory bound either.

I think my 'SA' and 'SB' implementation might have some inefficiencies.  Going to have a look.

**Comment by alreadydone** [comment]
However the `convolve3/sgemm` error persists at this moment. Why is it still called sgemm? I am not sure whether the tensor core tuning results go into XgemmBatched or XgemmBatchedHalf, but we need to delete leelaz_opencl_tuning and re-tune for tensor cores, right?

**Comment by ihavnoid** [comment]
The tensor core goes to XgemmBatchedHalf.  Tensor cores are for half precision only hence it skips trying to use tensor cores for single precision.

**Comment by alreadydone** [comment]
People tested using my https://github.com/alreadydone/lz/tree/fastexit-tensorcore merger of batching and tensor core support (all with 256x40 network, pos/s doesn't count NNCache hits while n/s counts them):
V100 --batchsize 12 : [before tuner](https://github.com/gcp/leela-zero/files/2612778/alreadydone-fastexittensorcore.before.tuner.txt) | [full tuner](https://github.com/gcp/leela-zero/files/2612779/alreadydone-fastexittensorcore.full-tuner.txt) | [after tuner](https://github.com/gcp/leela-zero/files/2612780/alreadydone-fastexittensorcore.after.full-tuner.txt) : 764 pos/s/GPU
V100 [full-tuner](https://github.com/gcp/leela-zero/files/2612791/-full-tuner.txt) (this I believe is batchsize 1)

GTX 1060 batchsize unknown : 10% increase
GTX 1070 --batchsize 12 : usual tuning, improves from 270 to 320 pos/s (~18.5% increase ! )
GTX 1080Ti  batchsize unknown : 20-30% increase (!) however another 2080Ti + 1080Ti x 2 setup reports only 8-9% increase (the 2080Ti is affected by #2040 though)
I have no idea why these cards without tensor cores get improvements! Maybe it's due to utilizing CUDA?

**Comment by ihavnoid** [comment]
@alreadydone it depends on what your base version was.  There was a couple of opencl optimization patches before the tensorcore patches.

**Comment by alreadydone** [comment]
I think most of them are comparing with my fastexit branch (formerly 1t-batch-fastexit).

**Comment by alreadydone** [comment]
The 764 pos/s/GPU on V100 x 2 I reported above is actually 1528 pos/s on a single V100. Somehow the tester was able to run the command with --gpu 0 --gpu 1 (~20% gain compared to without --gpu).

**Comment by alreadydone** [comment]
2080Ti, --batchsize 16, full tuner : ~1000 pos/s 
(using latest driver 416.94, surprisingly passed tuning without any huge error)
<details>

![image](https://user-images.githubusercontent.com/3064145/48981143-d5aab180-f09f-11e8-89e9-6ce839a8a57c.png)
</details>

I think some improvements on 10xx cards are due to your work #2014 and @Ttl's #2023, but that doesn't explain a 10+% improvement. Maybe full tuner?

**Comment by ihavnoid** [comment]
Okay.  Seems to provide some gain and at least doesn't crash immediately.  I will be cleaning up some of the code (e.g., removing hard-coded values, add more tuning parameters) and then push a PR for others to review.  Might take a couple of days.

**Comment by ihavnoid** [comment]
@already I don't think that's how the SM work.  It's not like we have separate instruction schedulers to do that.  Even if they do, we don't have enough parallelism to even saturate the tensor cores at this moment.  Based on my experiments we run out of power and DRAM bandwidth even on extremely large impractical batch sizes.

**Comment by ihavnoid** [comment]
No idea.  Maybe a full log might help, including the tuning results?

**Comment by ihavnoid** [comment]
One thing I know right know is that things go wrong when you revert to an older version of leela zero because the tuning results are not backward compatible.  I am thinking that this probably need some special notation on the tuner result to mark it as a result from a system with a tensor core.

**Comment by ihavnoid** [comment]
@alreadydone I pushed another commit.  This bumps up the tuner version so that the old versions don't confuse itself with the newer tuner results.  Please try after removing all prior tuning results to avoid any unexpected side-effects.

**Comment by ihavnoid** [comment]
I tested my code to make sure that it doesn't lose any strength due to additional computation errors.  A 60-game, `-v 200 --noponder` game with latest net (vs current next branch) resulted in 33-27 for the tensorcore implementation.  So it seems that we aren't losing any strength on equal playout.

**Comment by alreadydone** [comment]
After your amazing work it seems that we are already much faster than ELF on V100. 
From https://zhuanlan.zhihu.com/p/36353764 and https://zhuanlan.zhihu.com/p/36438750 I infer that ELF evaluates 80k positions in 50 seconds (at least at the time it played with Korean champions) , which amounts to 1600 n/s with batch size 16, while we are now hitting 3950 pos/s running --benchmark on Google Cloud with --batchsize 12 (with alreadydone/fastexit-tensor-accum+).

**Comment by alreadydone** [comment]
Mostly cards with tensor cores (Volta or Turing microarchitecture). See [a list](https://github.com/ihavnoid/hgemmtest#hgemmtest) here.

**Comment by alreadydone** [comment]
Though ELF(224x20) runs at ~4000 pos/s on V100, it only runs at ~2700 pos/s on Titan V. What could be the cause? Memory? So performance on V100 is close to 1st-gen TPU, but the AZ paper claims 1st-gen TPU is close to Titan V.

**Comment by ihavnoid** [comment]
No idea.  Can you test it with netbench?

**Comment by alreadydone** [comment]
netbench was broken since asynchronous NN eval was implemented, and I haven't been sure how to reimplement it (e.g. whether to include post-processing of NN outputs by the CPU). In the meantime, pos/s should be a good indicator as it directly measures the amount of computation done by the GPU per second.
-------
It has been reported that [these two lines](https://github.com/ihavnoid/leela-zero/commit/fa8b52cc9f852503c5ceee825c0ce6e8bdb6721e#diff-91dc351f9740337f63c56149696b3da9R31-R32) cause the engine to report `Tensor Core Support: No` for 2080Ti. Maybe it's just one of them but I haven't figured out which.

**Comment by ihavnoid** [comment]
@alreadydone can you check the driver version?

**Comment by alreadydone** [comment]
@ihavnoid There are multiple people who reported the problem and one of them is using 410.78 on Linux.

**Comment by ihavnoid** [comment]
@alreadydone okay so I removed the last two lines - it won't hurt not to test them because m16n16k16 is all we need to at least use tensor cores.  Please give it a try.

**Comment by alreadydone** [comment]
Yes I commented out those two line and they find that tensor core support is back. That's how I located the problem.
